File: Autonomous Mobile Robots/ex6_place_recognition_line_fitting_solution/matlab/line_regression.m
%% Load test data
clear;
clc;

x = readmatrix('x.csv');
y = readmatrix('y.csv');

draw_points(x, y);

N = size(x, 1); 
assert(N == size(y, 1));

% Compute the centroid of the point set (xmw, ymw) considering that
% the centroid of a finite set of points can be computed as
% the arithmetic mean of each coordinate of the points.
xc = sum(x) / N;
yc = sum(y) / N;

%% TODO: Part 1: Derive the parameter alpha

...

% compute alpha
alpha = ...;

%% Part 2: Derive the parameter r

% compute parameter r by inserting the centroid 
% into the line equation and solve for r
r = ...;

%% Draw the fitted line

hold off;
draw_points(x,y);
hold on;
draw_line(alpha, r);
hold off;

%% Utility functions

function draw_points(x, y)
scatter(x,y);
grid on;
xlabel('x');
ylabel('y');
end

function draw_line(alpha, r)
[xz, yz] = pol2cart(alpha, r);
line_origin = [xz; yz];
delta_vector = [ -sin(alpha); cos(alpha)] * 10;
line_points = [line_origin, line_origin + delta_vector];
line(line_points(1, :), line_points(2, :), 'color', 'r', 'linestyle', '-', 'linewidth', 2);
end

File: Autonomous Mobile Robots/ex6_place_recognition_line_fitting_solution/python/line_regression.py
import numpy as np
import matplotlib.pyplot as plt

def load_data(x_path, y_path):
    x=np.load(x_path)
    y=np.load(y_path)
    return x, y

def derive_alpha(x, y, xc, yc):
    # TODO: Implement this function
    alpha = ...
    return alpha

def derive_r(xc, yc, alpha):
    # TODO: Implement this function
    r = ...
    return r

def pol2cart(alpha, r):
    xz = r * np.cos(alpha)
    yz = r * np.sin(alpha)
    return xz, yz

def plot_fit(x, y, alpha, r):
    # Plot the points
    plt.scatter(x, y)

    # Plot the line
    xz, yz = pol2cart(alpha, r)
    line_foot_point = np.array([xz, yz])
    delta_vec = np.array([-np.sin(alpha), np.cos(alpha)]) * 10
    line_points = np.vstack((line_foot_point, line_foot_point + delta_vec))
    plt.plot(line_points[:,0], line_points[:, 1], color='red')

    plt.show()

if __name__ == '__main__':
    # Load test data
    x_path = './x.npy'
    y_path = './y.npy'
    x, y = load_data(x_path, y_path)

    # Compute the centroid
    N = x.shape[0]
    assert (N == y.shape[0])
    xc = np.sum(x) / N
    yc = np.sum(y) / N

    # Part 1: Derive the parameter alpha
    alpha = derive_alpha(x, y, xc, yc)

    # Part 2: Derive the parameter r using alpha
    r = derive_r(xc, yc, alpha)

    print(f'Found line with parameters alpha = {alpha} and r = {r}')

    plot_fit(x, y, alpha, r)
